sdk/src/azure/core/az_mqtt5_request_hfsm.c

```diff
@@ -0,0 +1,681 @@
+          this_policy->_internal.correlation_id);
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+/**
+ * @brief Failure state - locks up all execution of this policy
+ */
+static az_result faulted(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  az_mqtt5_request* this_policy = (az_mqtt5_request*)me;
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+    {
+      log_event(
+          event,
+          AZ_SPAN_FROM_STR("az_mqtt5_request_policy/faulted"),
+          this_policy->_internal.correlation_id);
+      _request_stop_timer(&this_policy->_internal.request_completion_timer);
+      /* we always send an error before transitioning to faulted, so no need to send another one
+      here, just return AZ_OK to confirm the transition was successful. Exception for
+      AZ_MQTT5_EVENT_REQUEST_FAULTED because an error is sent from the
+      rpc client when this event is sent*/
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUBACK_RSP:
+    case AZ_MQTT5_EVENT_SUBACK_RSP:
+    case AZ_MQTT5_EVENT_CONNECT_RSP:
+    case AZ_MQTT5_EVENT_DISCONNECT_RSP:
+    case AZ_MQTT5_EVENT_UNSUBACK_RSP:
+    case AZ_MQTT5_EVENT_PUB_RECV_IND:
+    case AZ_HFSM_EVENT_TIMEOUT:
+      // ignore, not from application
+      break;
+
+    case AZ_MQTT5_EVENT_REQUEST_COMPLETE:
+    case AZ_MQTT5_EVENT_REQUEST_FAULTED:
+    {
+      az_span* correlation_id = (az_span*)event.data;
+      if (az_span_is_content_equal(*correlation_id, this_policy->_internal.correlation_id))
+      {
+        log_event(
+            event,
+            AZ_SPAN_FROM_STR("az_mqtt5_request_policy/faulted"),
+            this_policy->_internal.correlation_id);
+        ret = AZ_ERROR_HFSM_INVALID_STATE;
+      }
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_RPC_CLIENT_REMOVE_REQ:
+    {
+      az_mqtt5_rpc_client_remove_req_event_data* event_data
+          = (az_mqtt5_rpc_client_remove_req_event_data*)event.data;
+      // Make sure this remove request is either for this request or any faulted request
+      if (az_span_is_content_equal(*event_data->correlation_id, AZ_SPAN_EMPTY)
+          || az_span_is_content_equal(
+              *event_data->correlation_id, this_policy->_internal.correlation_id))
+      {
+        log_event(
+            event,
+            AZ_SPAN_FROM_STR("az_mqtt5_request_policy/faulted"),
+            this_policy->_internal.correlation_id);
+        *event_data->correlation_id = this_policy->_internal.correlation_id;
+        *event_data->policy = this_policy;
+      }
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_REQUEST_INIT:
+      // ignore
+      break;
+
+    default:
+      log_event(
+          event,
+          AZ_SPAN_FROM_STR("az_mqtt5_request_policy/faulted"),
+          this_policy->_internal.correlation_id);
+      ret = AZ_ERROR_HFSM_INVALID_STATE;
+      break;
+  }
+
+  return ret;
+}
+
+AZ_NODISCARD az_result _az_mqtt5_request_hfsm_policy_init(
```

Comment:
This is our current pattern to separate the hfsm policy initialization from the larger initialization, but I think I agree that this might not be necessary

Reply:
removed across the HFSMs!